### ISSUE TITLE
Checkstyle hangs freezing intelliJ

### ISSUE THREAD

**Issue by thomas-klambauer** [jshiell/checkstyle-idea#515]
Had two projects windows open (completely different folders)
Ran checkstyle in one (open file -> scan module).
Saw already the results (thought scan was done?).
Closed that intellij project window where the scan was executed.
When looking at the other project - it was frozen - not recovering.

See attached screenshot and stacktrace in that state. Suspecting checkstyle.

**Comment by thomas-klambauer** [comment]
[intellijHang.txt](https://github.com/jshiell/checkstyle-idea/files/5964998/intellijHang.txt)
![OnPaste 20210211-125513](https://user-images.githubusercontent.com/12692832/107633475-5f86cb00-6c68-11eb-857d-9a7f57ba5329.png)

**Comment by jshiell** [comment]
Thanks for the report. After a quick look at the scan, I can't argue - it looks like it's stuck on the Checker cleanup. I'll try and have a look in the not too distant, but I fear it won't be this weekend thanks to illness.

```
"AWT-EventQueue-0" #27 prio=6 os_prio=0 cpu=511689,86ms elapsed=17050,16s tid=0x00007f55480dd800 nid=0x5c9c waiting for monitor entry  [0x00007f553a2e8000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.infernus.idea.checkstyle.service.cmd.OpDestroyChecker.execute(OpDestroyChecker.java:32)
	- waiting to lock <0x00000006dcbe9520> (a com.puppycrawl.tools.checkstyle.Checker)
	at org.infernus.idea.checkstyle.service.cmd.OpDestroyChecker.execute(OpDestroyChecker.java:15)
	at org.infernus.idea.checkstyle.service.CheckstyleActionsImpl.executeCommand(CheckstyleActionsImpl.java:130)
	at org.infernus.idea.checkstyle.service.CheckstyleActionsImpl.destroyChecker(CheckstyleActionsImpl.java:66)
	at org.infernus.idea.checkstyle.checker.CheckStyleChecker.destroy(CheckStyleChecker.java:47)
	at org.infernus.idea.checkstyle.checker.CachedChecker.destroy(CachedChecker.java:48)
	at org.infernus.idea.checkstyle.checker.CheckerFactoryCache.destroyChecker(CheckerFactoryCache.java:62)
	at org.infernus.idea.checkstyle.checker.CheckerFactoryCache$$Lambda$10977/0x00000008055e8440.accept(Unknown Source)
	at java.util.ArrayList.forEach(java.base@11.0.9.1/ArrayList.java:1541)
	at org.infernus.idea.checkstyle.checker.CheckerFactoryCache.invalidate(CheckerFactoryCache.java:56)
	at org.infernus.idea.checkstyle.checker.CheckerFactoryCache.dispose(CheckerFactoryCache.java:41)
	at com.intellij.openapi.util.ObjectTree.runWithTrace(ObjectTree.java:138)
	at com.intellij.openapi.util.ObjectTree.executeAll(ObjectTree.java:168)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:142)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:130)
	at com.intellij.serviceContainer.ComponentManagerImpl.dispose(ComponentManagerImpl.kt:952)
	at com.intellij.openapi.project.impl.ProjectExImpl.dispose(ProjectExImpl.kt:264)
	- locked <0x00000006233d3d30> (a com.intellij.openapi.project.impl.ProjectExImpl)
```

```

"ApplicationImpl pooled thread 273" #907 prio=4 os_prio=0 cpu=77956,09ms elapsed=484,98s tid=0x00007f54d02ae800 nid=0x17ff runnable  [0x00007f54904fb000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.9.1/Native Method)
	- parking to wait for  <0x00000005bdb87100> (a com.intellij.openapi.application.impl.ReadMostlyRWLock)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@11.0.9.1/LockSupport.java:234)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.waitABit(ReadMostlyRWLock.java:138)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.readLock(ReadMostlyRWLock.java:129)
	at com.intellij.openapi.application.impl.ApplicationImpl.acquireReadLock(ApplicationImpl.java:897)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:858)
	at org.infernus.idea.checkstyle.service.CheckStyleAuditListener.auditFinished(CheckStyleAuditListener.java:60)
	at com.puppycrawl.tools.checkstyle.Checker.fireAuditFinished(Checker.java:271)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:231)
	at org.infernus.idea.checkstyle.service.cmd.OpScan.processAndAudit(OpScan.java:76)
	- locked <0x00000006dcbe9520> (a com.puppycrawl.tools.checkstyle.Checker)
	at org.infernus.idea.checkstyle.service.cmd.OpScan.execute(OpScan.java:52)
```

**Comment by thomas-klambauer** [comment]
Thanks, don't sweat it, only saw this once. Get well soon!

**Comment by jshiell** [comment]
Thanks 😃 

This is fun. It looks like the scan has the lock on the Checker, however it needs read-lock from IDEA to get the scan results in the audit listener. However, a window-closing event has fired, which has made IDEA grab the write-lock (blocking the read-lock), and our cleanup tries to get the Checker lock ... which is of course waiting on the read-lock. Welp. 

Ideally we'd just time-out on the read of the scan results but IDEA's API, bless it, doesn't seem to offer any option to time-out. So I guess we'll need to timeout on the Checker lock ... which is currently done with `synchronised`, so means switching to a `Lock`. And `Checker` is a Checkstyle class, so we can't just add a lock member. God, I love concurrency.

TL;DR I know why; just need to think a bit before I start fixing it, as concurrency makes fools of us all.

**Comment by jshiell** [comment]
Hopefully fixed 🤞 I'll hold off on a release for a day or two, as I'm expecting Checkstyle 8.41 to be relased any day now.

**Comment by thomas-klambauer** [comment]
Thx!
